File: java/03-advanced/04-design-patterns/ObserverPattern.java
import java.util.*;

/**
 * LESSON: Observer Design Pattern
 * Defines a one-to-many dependency. When the subject changes state,
 * all registered observers are notified automatically.
 * Use cases: Event systems, UI updates, notifications.
 */
public class ObserverPattern {
    public static void main(String[] args) {
        // Create the subject (event manager)
        EventManager eventManager = new EventManager();

        // Create observers (listeners)
        EmailNotifier emailNotifier = new EmailNotifier("admin@example.com");
        LogNotifier logNotifier = new LogNotifier("app.log");
        SMSNotifier smsNotifier = new SMSNotifier("+1234567890");

        // Register observers for events
        eventManager.subscribe("order_placed", emailNotifier);
        eventManager.subscribe("order_placed", logNotifier);
        eventManager.subscribe("payment_received", emailNotifier);
        eventManager.subscribe("payment_received", smsNotifier);

        // Trigger events - all registered observers get notified
        System.out.println("===== Order Placed =====");
        eventManager.notify("order_placed", "Order #1234 placed by Emmanuel");

        System.out.println("\n===== Payment Received =====");
        eventManager.notify("payment_received", "Payment of $99.99 received");

        // Unsubscribe
        System.out.println("\n===== After unsubscribing email from order_placed =====");
        eventManager.unsubscribe("order_placed", emailNotifier);
        eventManager.notify("order_placed", "Order #1235 placed by Alice");
    }
}

// Observer interface
interface EventListener {
    void onEvent(String eventType, String data);
}

// Subject (Publisher) - manages observers
class EventManager {
    private final Map<String, List<EventListener>> listeners = new HashMap<>();

    void subscribe(String eventType, EventListener listener) {
        listeners.computeIfAbsent(eventType, k -> new ArrayList<>()).add(listener);
    }

    void unsubscribe(String eventType, EventListener listener) {
        List<EventListener> list = listeners.get(eventType);
        if (list != null) list.remove(listener);
    }

    void notify(String eventType, String data) {
        List<EventListener> list = listeners.get(eventType);
        if (list != null) {
            for (EventListener listener : list) {
                listener.onEvent(eventType, data);
            }
        }
    }
}

// Concrete Observers
class EmailNotifier implements EventListener {
    private final String email;
    EmailNotifier(String email) { this.email = email; }

    @Override
    public void onEvent(String eventType, String data) {
        System.out.printf("[EMAIL -> %s] Event: %s | Data: %s%n", email, eventType, data);
    }
}

class LogNotifier implements EventListener {
    private final String logFile;
    LogNotifier(String logFile) { this.logFile = logFile; }

    @Override
    public void onEvent(String eventType, String data) {
        System.out.printf("[LOG -> %s] Event: %s | Data: %s%n", logFile, eventType, data);
    }
}

class SMSNotifier implements EventListener {
    private final String phone;
    SMSNotifier(String phone) { this.phone = phone; }

    @Override
    public void onEvent(String eventType, String data) {
        System.out.printf("[SMS -> %s] Event: %s | Data: %s%n", phone, eventType, data);
    }
}

File: cpp-practice/03-advanced/03-design-patterns/observer_pattern.cpp
/**
 * LESSON: Observer Pattern
 * Defines a one-to-many dependency: when one object (Subject) changes state,
 * all its dependents (Observers) are notified automatically.
 *
 * Compile: g++ -std=c++17 -o observer observer_pattern.cpp
 * Run:     ./observer
 */
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <memory>
#include <functional>

// ===== CLASSIC OBSERVER PATTERN =====

// Observer interface
class IObserver {
public:
    virtual ~IObserver() = default;
    virtual void update(const std::string& event, const std::string& data) = 0;
};

// Subject (Observable) base class
class Subject {
    std::vector<IObserver*> observers;

public:
    virtual ~Subject() = default;

    void subscribe(IObserver* obs) {
        observers.push_back(obs);
    }

    void unsubscribe(IObserver* obs) {
        observers.erase(
            std::remove(observers.begin(), observers.end(), obs),
            observers.end());
    }

    void notify(const std::string& event, const std::string& data) {
        for (auto* obs : observers) {
            obs->update(event, data);
        }
    }
};

// ===== WEATHER STATION EXAMPLE =====
class WeatherStation : public Subject {
    double temperature = 0;
    double humidity = 0;

public:
    void setMeasurements(double temp, double hum) {
        temperature = temp;
        humidity = hum;
        std::cout << "\n[WeatherStation] New data: " << temp << "°C, "
                  << hum << "% humidity\n";
        notify("weather_update", std::to_string(temp) + "," + std::to_string(hum));
    }
};

class PhoneDisplay : public IObserver {
    std::string owner;
public:
    PhoneDisplay(const std::string& o) : owner(o) {}
    void update(const std::string& event, const std::string& data) override {
        std::cout << "  [" << owner << "'s Phone] " << event << ": " << data << "\n";
    }
};

class WebDashboard : public IObserver {
public:
    void update(const std::string& event, const std::string& data) override {
        std::cout << "  [WebDashboard] Updating chart with: " << data << "\n";
    }
};

class AlertSystem : public IObserver {
public:
    void update(const std::string& event, const std::string& data) override {
        double temp = std::stod(data.substr(0, data.find(',')));
        if (temp > 35.0) {
            std::cout << "  [ALERT] High temperature warning: " << temp << "°C!\n";
        }
    }
};

// ===== MODERN APPROACH: Callback-based =====
class EventEmitter {
    using Callback = std::function<void(const std::string&)>;
    std::vector<std::pair<std::string, Callback>> listeners;
    int nextId = 0;

public:
    int on(const std::string& event, Callback cb) {
        listeners.push_back({event, std::move(cb)});
        return nextId++;
    }

    void emit(const std::string& event, const std::string& data) {
        for (const auto& [evt, cb] : listeners) {
            if (evt == event) cb(data);
        }
    }
};

int main() {
    // --- Classic Observer ---
    std::cout << "=== Classic Observer Pattern ===\n";
    WeatherStation station;

    PhoneDisplay phone1("Emmanuel");
    PhoneDisplay phone2("Alice");
    WebDashboard dashboard;
    AlertSystem alert;

    station.subscribe(&phone1);
    station.subscribe(&phone2);
    station.subscribe(&dashboard);
    station.subscribe(&alert);

    station.setMeasurements(28.5, 65.0);
    station.setMeasurements(36.2, 70.0);  // triggers alert!

    // Unsubscribe Alice
    std::cout << "\n[Unsubscribing Alice...]\n";
    station.unsubscribe(&phone2);
    station.setMeasurements(22.0, 55.0);

    // --- Modern callback-based ---
    std::cout << "\n=== Modern Callback-Based ===\n";
    EventEmitter emitter;

    emitter.on("click", [](const std::string& data) {
        std::cout << "  Button clicked: " << data << "\n";
    });

    emitter.on("click", [](const std::string& data) {
        std::cout << "  Analytics: click on " << data << "\n";
    });

    emitter.on("hover", [](const std::string& data) {
        std::cout << "  Tooltip for: " << data << "\n";
    });

    emitter.emit("click", "Submit Button");
    emitter.emit("hover", "Help Icon");
    emitter.emit("click", "Cancel Button");

    return 0;
}
